libsql-sqlite3/src/vectordiskann.c

```diff
@@ -0,0 +1,345 @@
+    sqlite3_blob_close(pBlobSpot->pBlob);
+  }
+  if( pBlobSpot->pBuffer != NULL ){
+    sqlite3_free(pBlobSpot->pBuffer);
+  }
+  sqlite3_free(pBlobSpot);
+}
+
+/**************************************************************************
+** Layout specific utilities
+**************************************************************************/
+
+int nodeEdgesMaxCount(const DiskAnnIndex *pIndex){
+  unsigned int nMaxEdges = (pIndex->nBlockSize - pIndex->nNodeVectorSize - VECTOR_NODE_METADATA_SIZE) / (pIndex->nEdgeVectorSize + VECTOR_EDGE_METADATA_SIZE);
+  assert( nMaxEdges > 0);
+  return nMaxEdges;
+}
+
+int nodeEdgesMetadataOffset(const DiskAnnIndex *pIndex){
+  unsigned int offset;
+  unsigned int nMaxEdges = nodeEdgesMaxCount(pIndex);
+  offset = VECTOR_NODE_METADATA_SIZE + pIndex->nNodeVectorSize + nMaxEdges * pIndex->nEdgeVectorSize;
+  assert( offset <= pIndex->nBlockSize );
+  return offset;
+}
+
+void nodeBinInit(const DiskAnnIndex *pIndex, BlobSpot *pBlobSpot, u64 nRowid, Vector *pVector){
+  assert( VECTOR_NODE_METADATA_SIZE + pIndex->nNodeVectorSize <= pBlobSpot->nBufferSize );
+
+  memset(pBlobSpot->pBuffer, 0, pBlobSpot->nBufferSize);
+  writeLE64(pBlobSpot->pBuffer, nRowid);
+  writeLE16(pBlobSpot->pBuffer + sizeof(u64), 0);
+  vectorSerializeToBlob(pVector, pBlobSpot->pBuffer + VECTOR_NODE_METADATA_SIZE, pIndex->nNodeVectorSize);
+}
+
+void nodeBinVector(const DiskAnnIndex *pIndex, const BlobSpot *pBlobSpot, Vector *pVector) {
+  assert( VECTOR_NODE_METADATA_SIZE + pIndex->nNodeVectorSize <= pBlobSpot->nBufferSize );
+
+  vectorInitStatic(pVector, pIndex->nNodeVectorType, pBlobSpot->pBuffer + VECTOR_NODE_METADATA_SIZE, pIndex->nNodeVectorSize);
+}
+
+u16 nodeBinEdges(const DiskAnnIndex *pIndex, const BlobSpot *pBlobSpot) {
+  assert( VECTOR_NODE_METADATA_SIZE <= pBlobSpot->nBufferSize );
+
+  return readLE16(pBlobSpot->pBuffer + sizeof(u64));
+}
+
+void nodeBinEdge(const DiskAnnIndex *pIndex, const BlobSpot *pBlobSpot, int iEdge, u64 *pRowid, Vector *pVector) {
+  int offset = nodeEdgesMetadataOffset(pIndex);
+
+  if( pRowid != NULL ){
+    assert( offset + (iEdge + 1) * VECTOR_EDGE_METADATA_SIZE <= pBlobSpot->nBufferSize );
+    *pRowid = readLE64(pBlobSpot->pBuffer + offset + iEdge * VECTOR_EDGE_METADATA_SIZE + sizeof(u64));
+  }
+  if( pVector != NULL ){
+    assert( VECTOR_NODE_METADATA_SIZE + pIndex->nNodeVectorSize + iEdge * pIndex->nEdgeVectorSize < offset );
+    vectorInitStatic(
+      pVector,
+      pIndex->nEdgeVectorType,
+      pBlobSpot->pBuffer + VECTOR_NODE_METADATA_SIZE + (iEdge + 1) * pIndex->nNodeVectorSize,
+      pIndex->nEdgeVectorSize
+    );
+  }
+}
+
+int nodeBinEdgeFindIdx(const DiskAnnIndex *pIndex, const BlobSpot *pBlobSpot, u64 nRowid) {
+  int i, nEdges = nodeBinEdges(pIndex, pBlobSpot);
+
+  for(i = 0; i < nEdges; i++){
+    u64 edgeId;
+    nodeBinEdge(pIndex, pBlobSpot, i, &edgeId, NULL);
+    if( edgeId == nRowid ){
+      return i;
+    }
+  }
+  return -1;
+}
```

Comment:
Just an idea for future improvements. If edges are stored in sorted order then we could use binary search here.

Reply:
It might be counter productive with number of edges <= 2^16 though

Reply:
Yes, nice suggestion. I will add todo here.

I think the difference will became visible earlier (I guess at 1024 elements bin search must be faster) - but the main question here is in the real bottlenecks. We will measure performance and start optimization from most heavy places and I think this is not the top perf eater.